README.md
# discord-bot-honger
Discord bot for doing misc related activity


CHANGELOG.md
# Changelog
## [1.2.3] -2017-8-30
## Added
- Selectable channel for Hong'er reply

## [1.2.2] -2017-8-29
## Added
- `onSuccess` to return custom message

## Changed
- broadcast custom message

## [1.2.1b] -2017-8-28
## Added
- `pollingTime` to log last polling time
- Timeout to axios request

## Changed
- Fix polling schedule not invoked in case of timeout

## [1.2.1] -2017-8-25
## Added
- `parseData()` to polling object
- `htl` polling

## Changed
- Readjust min `!timeout` to 20 seconds

## Removed
- `render()`, now automatically title + url
- `check()`, now use universal checker

## [1.2.0] -2017-8-22
### Added
- `CHANGELOG.md`
- `!stats` to check current bot's status
- `!poll <on/off>` to turn on/off polling state
- `!check` alias naming
- `watch` option for NPM

### Changed
- Fix `!timeout <NEW_TIMER>` to use second instead of millisecond
- Fix `!check` to show all pollings by default
- Improve `!check` display message
- Improve `lnmtl` data efficiency
- Improve polling class structure

### Removed
- `!timeout` check command, please use `!stats` instead

main.js
const _ = require('lodash');
const Discord = require('discord.js');
const AsyncPolling = require('async-polling');
const cheerio = require('cheerio');
const axios = require('./hyper.js');
const http = require('http');
const math = require('mathjs');
const moment = require('moment-timezone');
const command = require('./command.js');
const pollingList = require('./polling.js');
const pkg = require('./package.json');
const config = require('./config.test.json');

require('v8-profiler');

moment.tz.setDefault('Asia/Shanghai');

class Bot {
    constructor() {
        this.client = new Discord.Client();
        this.pollingList = pollingList;

        // create pollings
        _.forOwn(this.pollingList, (o, k) => {
            let delay = _.isNumber(o.delay) ? o.delay : _.isFunction(o.delay) ? o.delay() : null;
            let promise = null;
            o.asyncPolling = AsyncPolling(end => {
                // save current time
                this.lastPollingTime = moment();

                if(_.isArray(o.url)) {
                    let all = o.url.map((url, i) => {
                        return axios.get(url, o.axiosOptions)
                            .then( res=> {
                                return o.parseData[i](this, res.data);
                            });
                    });

                    promise = axios.all(all)
                        .then(res => {
                            if(!o.latestChapter) {
                                o.latestChapter = res[0] || res[1];
                                o.onLoadChapter(this);
                                return null;
                            } else {
                                let newData = null;
                                res.forEach(e => {
                                    if(e && e.title !== o.latestChapter.title) {
                                        newData = e;
                                    }
                                });
                                if(newData) {
                                    o.latestChapter = newData;
                                }
                                return newData ? o : null;
                            }
                        }, e => {
                            console.error(e);
                            end();
                        });
                }
                else {
                    promise = axios.get(o.url, o.axiosOptions)
                        .then(res => {
                            // return null if you don't want it to reports.
                            let newData = o.parseData(this, res.data);
                            if(!o.latestChapter) {
                                // first time
                                o.latestChapter = newData;
                                o.onLoadChapter(this);
                                return null;
                            } else {
                                if(o.latestChapter.url !== newData.url) {
                                    o.latestChapter = newData;
                                    return o;
                                }
                                return null;
                            }
                        }, e => {
                            //timeout constantly...
                            
                            console.error(e);
                            end();
                        });
                }

                promise.then(o => {
                    if(_.isObject(o)) {
                        let str = o.onNewChapter(this);
                        this.broadcast(str, 'notices');
                    }
                    end();
                }, e => {
                    console.log('95')
                    console.error(e);
                    end();
                })
                .catch(e => {
                    console.log('99')
                    console.error(e);
                    end();
                });
           },  delay);
        });

        this.changeDelayInterval = AsyncPolling(end => {
            _.forOwn(this.pollingList, o => {
                if(_.isFunction(o.delay)) {
                    let newDelay = o.delay();
                    if(newDelay !== o.asyncPolling._delay) {
                        o.asyncPolling._delay = newDelay;
                    }
                }
            });
        }, 60000 * 10);

        // on ready
        this.client.on('ready', () => {
            console.log('Bot is ready');
            console.log(`https://discordapp.com/api/oauth2/authorize?client_id=${config.client_id}&scope=bot&permissions=${config.permissions}`);
    
            this.startTime = moment();
            this.lastPollingTime = moment();
            this.changeDelayInterval.run();
            this.runPoll();
            this.broadcast('Hong\'er v' + pkg.version + ' is ready!!\nOhhh!!!');
        });
        
        // msg listener
        this.client.on('message', message => {
            // only this channel
            if(message.channel.name !== 'talk-to-honger') {
                return;
            }

            // only if Hong'er is mentioned
            if(!message.mentions.everyone && message.mentions.users.find('id', this.client.user.id)) {
                message.channel.send(':sleeping:');
            }
        
            // command start with !
            if(message.content.startsWith('!')) {
                let token = message.content.split(' '); //match(/(?:[^\s"]+|"[^"]*")+/g) 
                let cmd = token[0].substr(1).toLowerCase();
                let args = token.slice(1);                
                let result = _.pickBy(command, (v, k) => _.startsWith(k, cmd));
                let resultKeys = _.keys(result);
                let done = null;

                if(cmd.length === 0) {
                }
                else if(resultKeys.length === 1) {
                    done = result[resultKeys[0]].action(this, message, ...args);
                } else if(resultKeys.length > 1) {
                    let exact = _.find(resultKeys, (k) => parseInt(k.split('_')[1]) === args.length);
                    
                    // exactly arg len
                    if(exact) {
                        done = result[exact].action(this, message, ...args);
                    } else {
                        // more than arg len
                        let more = _.reduce(resultKeys, (candidate, k) => {
                            let n = parseInt(k.split('_')[1]);
                            let c = parseInt(candidate.split('_')[1]);
                            if(c < n) {
                                return k;
                            }
                            return candidate;
                        }, resultKeys[0]);
                        
                        done = result[more].action(this, message, ...args);
                    }
                }

                // error handling
                if(done === false) {
                    message.channel.send('Error: wrong conditions or arguments');
                }
                if(done === null) {
                    //message.channel.send('Error: cannot find command');
                }
            }
        });
    }
    broadcast(content, name) {
        this.client.guilds.array().forEach(g => {
            let channel = g.channels.find('name', name || 'talk-to-honger');
            if(channel) {
                channel.send(content);
            }
        });
    }
    connect() {
        // dummy http server
        http.createServer(function(req,res) {
            res.writeHead(200);
            res.write('Launched!');
            res.end();
        }).listen(process.env.PORT || 3000);

        // connect to discord
        this.client.login(config.token);
    }
    runPoll(name) {
        if(name) {
            this.pollingList[name].asyncPolling.run()
        }
        else {
            _.forOwn(this.pollingList, e => e.asyncPolling ? e.asyncPolling.run() : null);
        }
    }
    stopPoll(name) {
        if(name) {
            this.pollingList[name].asyncPolling.stop();
        }
        else {
            _.forOwn(this.pollingList, e => e.asyncPolling ? e.asyncPolling.stop() : null);
        }
    }
}

new Bot().connect();

hyper.js
const hyperquest = require('hyperquest');
const _ = require('lodash');
const Promise = require('bluebird');
const url = 'https://m.zongheng.com/h5/ajax/chapter/list?h5=1&bookId=408586&pageNum=1&pageSize=1&chapterId=0&asc=1';
const iconv = require('iconv-lite');


var hyper = {

    get: function(url, opts) {
        opts = _.extend({
            encoding: 'utf8'
        }, opts);
        
        return new Promise(function(resolve, reject) {
            hyperquest(url, {}, function(err, res) {
                if(err) {
                    reject(err);
                    return;
                }
                var data = [];
                res.on('data', function(chunk) {
                    data.push(chunk);
                });

                res.on('end', function() {
                    let buffer = Buffer.concat(data);
                    let parse = buffer;

                    if(opts.encoding) {
                        parse = iconv.decode(parse, opts.encoding);
                    }
                    if(res.headers['content-type'].includes('application/json')) {
                        parse = JSON.parse(parse);
                    }

                    resolve({data: parse});
                });

                res.on('error', function(e) {
                    reject(e);
                });
            });
        });
    },

    all: function(arr) {
        return Promise.all(arr);
    }
}
module.exports = hyper;


command.js
const poll = require('./polling.js');
const Discord = require('discord.js');
const math = require('mathjs');
const moment = require('moment-timezone');
const _ = require('lodash');

moment.tz.setDefault('Asia/Shanghai');

var checkOptions = _.keys(poll);
var commands = {
  'reset': {
    description: 'Reset current chapter data, in case data are missing.',
    action: function(bot, msg) {
      bot.stopPoll();
      _.forOwn(bot.pollingList, v => {
        v.latestChapter = undefined;
      });
      bot.runPoll();
      msg.channel.send('reset completed');
    }
  },
  'status_0' : {
    description: 'Check bot status',
    action: function(bot, msg) {
      let content = '';
      let duration = moment.duration(moment().diff(bot.startTime));
      let pollDuration = moment.duration(moment().diff(bot.lastPollingTime));
      content += `Last Poll: **${Math.round(pollDuration.asSeconds())}** seconds ago\n`;
      content += `Running time: **${duration.hours()}**h **${duration.minutes()}**m **${duration.seconds()}**s\n`;
      _.forOwn(bot.pollingList, (v,k) => {
        content += k.toUpperCase() + ': **' + (v.asyncPolling._mustSchedule ? (v.asyncPolling._delay/1000 + '**s') : 'OFF**') + '\n';
      });
      msg.channel.send(content);
      return true;
    }
  },
  'check': {
    description: 'Check latest content from site',
    args: [checkOptions.join('|')],
    render: function(bot, obj) {
      return obj.title + '\n' + obj.url;
    },
    action: function(bot, msg, name) {
      let content = null;
      let o = null;
      
      // check all 
      if(_.has(poll, name)) {
        // normal render
        content = this.render(bot, bot.pollingList[name]);
      }
      else if((o = _.findIndex(checkOptions, e => e.startsWith(name)) ) >= 0) {
        // aliasing
        content = this.render(bot, bot.pollingList[checkOptions[o]]);
      }  
      else {
        // render everything
        let embed = new Discord.RichEmbed()
          .setTitle('Current Chapters')
          .setColor(0xc70000);
        checkOptions.forEach(name => {
          embed.addField(name.toUpperCase() + ' ' + (bot.pollingList[name].latestChapter || {}).title, (bot.pollingList[name].latestChapter || {}).url);
        });
        
        // render rich text
        content = {embed};
      }

      // no data?
      if(!content) {
        content = 'Error: cannot find data';
      }
      msg.channel.send(content);
      return true;
    }
  },
  'help': {
    description: 'List of all commands',
    action: function(bot, msg) {
      let cmdKeys = _.keys(commands).sort();
      let len = 0;
      let contents = _.map(cmdKeys, k => {
        let c = `**!${k.split('_')[0]}** ` + (commands[k].args ? _.map(commands[k].args, e => `<${e}>`).join(' ') : '');
        if(c.length > len) {
          len = c.length;
        }

        return c;
      });
      
      // formatting spaces
      let content = _.map(contents, (c, i) => {
        for(let i=0; i < len - c.length; i++) {
          c += '\t';
        }
        c += `\t\t\t${commands[cmdKeys[i]].description}`;
        return c;
      }).join('\n');
      msg.channel.send(content);
      return true;
    }
  },
  'test': {
    action: function(bot, msg, name) {
      bot.pollingList[name].latestChapter.title = 'something else';
      bot.pollingList[name].latestChapter.url = 'something else';
      console.log(bot.pollingList[name].latestChapter);
    }
  }
};

module.exports = commands;

polling.js
const _ = require('lodash');
const Discord = require('discord.js');
const moment = require('moment-timezone');
const cheerio = require('cheerio');

moment.tz.setDefault('Asia/Shanghai');

// get span object for checking between hour-time
function span(now) {
  today = moment(now).startOf('day');
  tmrow = moment(today).add(1, 'd');

  return function(from, to) {
    let f = moment(today);
    let t = moment(today);
    if(from > to) {
      t = moment(tmrow);
    }

    // prevent overlapping
    f.add(1, 'ms');

    f.hour(from);
    t.hour(to);

    return now.isBetween(f, t);
  }
}

// diff between two time as hour with decimals
function diff(a, b) {
  return moment.duration(a.diff(b)).asHours;
}

let pollings = {
  'raw': {
    alias: ['r'],
    url: 'https://m.zongheng.com/h5/ajax/chapter/list?h5=1&bookId=408586&pageNum=1&pageSize=1&chapterId=0&asc=1',
    parseData: function(bot, data) {
      // now is zongheng format
      return {
        title: data.chapterlist.chapters[0].chapterName,
        url: `http://book.zongheng.com/chapter/408586/${data.chapterlist.chapters[0].chapterId}.html`
      }
    },
    onNewChapter: function(bot) {
      bot.runPoll('free');
      return '@everyone, New Chapter from RAW!!!\n' + this.latestChapter.url;
    },
    delay: function() {
      let s = span(moment());

      if(s(22,24)) {
        return 5000;
      }
      if(s(0,4)) {
        return 5000;
      }
      return 15000;
    }
  },
  'free': {
    alias: ['f'],
    url: ['http://m.zwda.com/nitianxieshen/', 'http://www.piaotian.com/bookinfo/6/6760.html'],
    parseData: [function(bot, data) {
      let $ = cheerio.load(data);
      let tag = $('.block_txt2 > p').last().find('a');
      return {
        title: tag.text().trim(),
        url: 'http://www.zwda.com' + tag.attr('href')
      };
    }, function(bot, data) {
      let $ = cheerio.load(data);
      let tag = $('span.hottext:contains("最新章节：")').next();
      return {
        title: tag.text().trim(),
        url: tag.attr('href')
      };
    }],
    onLoadChapter: function(bot) {
      bot.stopPoll('free');
    },
    onNewChapter: function(bot) {
      bot.stopPoll('free');
      bot.runPoll('lnmtl');
      return '<@244072217623658506>, Free RAW is Up!!!\n' + this.latestChapter.url;
    },
    delay: 15000,
    axiosOptions: {
      encoding: 'gbk'
    }
  },
  'lnmtl': {
    alias: ['m', 'mtl'],
    url: 'https://lnmtl.com/novel/against-the-gods',
    parseData: function(bot, data) {
      let $ = cheerio.load(data);
      let newArray = [];

      // get list of all latest chapters on lnmtl page
      $('a.chapter-link').each(function() {
        let url = $(this).attr('href');
        let title = $(this).text().trim().replace('  ', ' ');
        let date = $(this).parent().next().find('span.label-default').text().trim();

        newArray.push({url, title, date});
      });

      // latest release
      let latest =  _.reverse(_.sortBy(newArray, 'date'))[0];
      let url = latest.url;
      let title = latest.title;
      return {url, title};
    },
    onLoadChapter: function(bot) {
      bot.stopPoll('lnmtl');
    },
    onNewChapter: function(bot) {
      bot.stopPoll('lnmtl');
      return '@everyone, LNMTL is up!!!\n' + this.latestChapter.url;
    },
    delay: 15000
  },
  'htl': {
    alias: ['h', 'ww', 'alyschu'],
    url: 'http://www.wuxiaworld.com/category/atg-chapter-release/',
    parseData: function(bot, data) {
      let $ = cheerio.load(data);
      let sel = $('article.category-atg-chapter-release').first();
      let url = sel.find('.entry-content a[href^="http://www.wuxiaworld.com/atg-index/"]').attr('href');
      let title = sel.find('.entry-title').text().trim();

      return {url, title};
    },
    onNewChapter: function(bot) {
      return 'New Chapter from HTL!!!\n' + this.latestChapter.url;
    },
    delay: 60000 * 5 //every 5 min
  }
};


// defaults
_.forOwn(pollings, (v,k) => {
  _.defaults(v, {
    onLoadChapter: function() {},
    onNewChapter: function(bot) {
      return '@everyone, New Chapter from ' + obj.name + '!!!\n' + this.latestChapter.url;
    },
    asyncPolling: null,
    lastChapterTime: null,
  });
});
module.exports = pollings;